### PR TITLE
perf: Optimize XLSX export by using more efficient iteration

### PR DESCRIPTION
**Pull request by alexander-yakushev**
When XLSX export is performed, a considerable amount of time is spent inefficiently iterating over lazy sequences to produce exported cells. This PR ensures that the iterated collections are constructed eagerly, and replaces `doseq` with Java iterators.

I also rewrote `rounds-to-int?` to not use bigdecimals as converting to it involves string parsing and allocations. The behavior for some numbers may change, such as `1.005` now rounds to `1.00`, not to `1.001` (because in double representation `1.005 * 100` is `100.49999...`). I can revert this change if this is important; the performance improvement from this is not critical.

### CODE REVIEW

File: src/metabase/query_processor/streaming/xlsx.clj [hunk 1]
```diff
@@ -290,9 +290,9 @@
 
 (defn- rounds-to-int?
   "Returns whether a number should be formatted as an integer after being rounded to 2 decimal places."
-  [value]
-  (let [rounded (.setScale (bigdec value) 2 java.math.RoundingMode/HALF_UP)]
-    (== (bigint rounded) rounded)))
+  [^double value]
+  (let [x100 (* value 100.0)]
+    (== (/ (Math/round x100) 100.0) (Math/floor value))))
```
Comment (crisptrutski) [inline src/metabase/query_processor/streaming/xlsx.clj:295]: This doesn't seem to be quite right. I think we need accurate precision at least for small numbers like this.

```clojure
(->> [1.006 1.005 1.004 -1.006 -1.005 -1.004 1.994 1.995 1.996 -1.994 -1.995 -1.996]
     (keep (fn [v]
             (let [old (let [rounded (.setScale (bigdec v) 2 java.math.RoundingMode/HALF_UP)]
                         (== (bigint rounded) rounded))
                   new (let [x100 (* v 100.0)]
                         (== (/ (Math/round x100) 100.0) (Math/floor v)))]
               (when (not= old new)
                 [v old new])))))

;; [-1.004 true false]
;; [1.995  true false]
;; [1.996  true false]
;; [-1.995 true false]
```

File: test/metabase/query_processor/streaming/xlsx_test.clj [hunk 2]
```diff
@@ -432,7 +432,7 @@
       (is (= [["#,##0"] ["#,##0.##"] ["#,##0"] ["#,##0.##"] ["#,##0"] ["#,##0.##"]]
              (rest (xlsx-export [{:field_ref [:field 0] :name "Col" :semantic_type :type/Cost}]
                                 {}
-                                [[1] [1.23] [1.004] [1.005] [10000000000] [10000000000.123]]
+                                [[1] [1.23] [1.004] [1.006] [10000000000] [10000000000.123]]
```
Comment (crisptrutski) [inline test/metabase/query_processor/streaming/xlsx_test.clj:435]: This seems like cheating 😄

**Review by crisptrutski — COMMENTED** [review]
Really nice, but perhaps let's leave the rounding optimization for now?

File: src/metabase/query_processor/streaming/xlsx.clj (hunk 1 above)
Comment (crisptrutski) [reply src/metabase/query_processor/streaming/xlsx.clj:295]: I see you called this out in the description 🤦 

I strongly suspect customers will be surprised by this formatting change. I also suspect there is another way to do this efficiently, perhaps using mod?

Comment (crisptrutski) [reply src/metabase/query_processor/streaming/xlsx.clj:295]: I tried out a whole bunch of variants of both float arithmetic and direct operations on the underlying bits, and there were always some cases I couldn't get right.

We can at least save conversion to a BigInt I think though:

```clojure
(defn rounds-to-int? [v]
  (let [bd     (BigDecimal/valueOf ^double v)
        scaled (.setScale bd 2 RoundingMode/HALF_UP)]
    (zero? (.scale (.stripTrailingZeros scaled)))))
```

Comment (crisptrutski) [reply src/metabase/query_processor/streaming/xlsx.clj:295]: Some optimizations I'm not sure would be worth it.

```clojure
(defn integer-when-rounded? [^double v]
  (or (= v (Math/round v))
      (let [bd (BigDecimal/valueOf v)]
        (or (< (.scale bd) 3)
            (not (pos? (-> bd (.setScale 2 RoundingMode/HALF_UP) .stripTrailingZeros .scale)))))))
```

**Comment by alexander-yakushev** [comment]
Reverted the `rounds-to-int?` change. There are perhaps other ways to rewrite it, but the impact does not justify the trouble for now.

**Review by crisptrutski — APPROVED** [review]

**Comment by crisptrutski** [comment]
@metabase-bot backport release-x.50.x

**Pull request merged**